homeassistant/components/media_player/cast.py

```diff
@@ -33,90 +38,164 @@
     vol.Optional(CONF_HOST): cv.string,
     vol.Optional(CONF_IGNORE_CEC): [cv.string],
 })
 
 
-# pylint: disable=unused-argument
-def setup_platform(hass, config, add_devices, discovery_info=None):
-    """Set up the cast platform."""
-    import pychromecast
+def _setup_internal_discovery(hass: HomeAssistantType) -> None:
+    """Set up the pychromecast internal discovery."""
+    if hass.data.get(INTERNAL_DISCOVERY_RUNNING_KEY, False):
+        # Internal discovery is already running
+        return
 
-    # Import CEC IGNORE attributes
-    pychromecast.IGNORE_CEC += config.get(CONF_IGNORE_CEC, [])
+    import pychromecast
 
-    known_hosts = hass.data.get(KNOWN_HOSTS_KEY)
-    if known_hosts is None:
-        known_hosts = hass.data[KNOWN_HOSTS_KEY] = []
+    def internal_callback(name):
+        """Called when zeroconf has discovered a new chromecast."""
+        mdns = listener.services[name]
+        ip_address, port, uuid, model_name, friendly_name = mdns
+        key = (ip_address, port, uuid)
 
-    if discovery_info:
-        host = (discovery_info.get('host'), discovery_info.get('port'))
+        if key in hass.data[KNOWN_CHROMECASTS_KEY]:
+            _LOGGER.debug("Discovered previous chromecast %s", mdns)
+            return
 
-        if host in known_hosts:
+        _LOGGER.debug("Discovered new chromecast %s", mdns)
+        try:
+            chromecast = pychromecast._get_chromecast_from_host(
+                mdns, blocking=True)
+        except pychromecast.ChromecastConnectionError:
+            _LOGGER.debug("Can't set up cast with mDNS info %s. "
+                          "Assuming it's not a Chromecast", mdns)
             return
+        hass.data[KNOWN_CHROMECASTS_KEY][key] = chromecast
+        dispatcher_send(hass, SIGNAL_CAST_DISCOVERED, chromecast)
 
-        hosts = [host]
+    _LOGGER.debug("Starting internal pychromecast discovery.")
+    listener, browser = pychromecast.start_discovery(internal_callback)
+    hass.data[INTERNAL_DISCOVERY_RUNNING_KEY] = True
 
-    elif CONF_HOST in config:
-        host = (config.get(CONF_HOST), DEFAULT_PORT)
+    def stop_discovery():
+        """Stops discovery of new chromecasts."""
+        pychromecast.stop_discovery(browser)
+        hass.data[INTERNAL_DISCOVERY_RUNNING_KEY] = False
 
-        if host in known_hosts:
-            return
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_discovery)
+
+
+@callback
+def _async_create_cast_device(hass, chromecast):
+    """Create a CastDevice Entity from the chromecast object.
 
-        hosts = [host]
+    Returns None if the cast device has already been added. Additionally,
+    automatically updates existing chromecast entities.
+    """
+
+    if chromecast.uuid is None:
+        # Found a cast without UUID, we don't store it because we won't be able
+        # to update it anyway.
+        return CastDevice(chromecast)
+
+    # Found a cast with UUID
+    known_casts = hass.data[ADDED_CAST_DEVICES_KEY]
```

Comment:
So you have "known" and "added" casts, and here you assign the `ADDED_CAST_DEVICES_KEY` stuff to `known_casts`. I'm confused. What's the difference. Should this be `added_casts = …`?

Reply:
Sure, I'm using both a bit interchangeably here since every time we "know" a chromecast, it is also added.